### PR TITLE
Fixed IQAE bug mentioned in issue 9280

### PR DESCRIPTION
**Pull request by shifubear**
### Summary
Reopening a PR due to account conflicts present in PR #9289. 

### Details and comments

Fixed the bug mentioned in issue #9280, and added more tests to increase robustness for phase estimation algorithms.

### CODE REVIEW

**Comment by shifubear** [comment]
Hi @woodsp-ibm, I reopened a new PR with the same changes here since there were some issues in the history of PR #9289 that seemed challenging to resolve.

File: releasenotes/notes/iterative-phase-estimation-bugfix-b676ffc23cea8251.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    Fixed a bug where the iterative phase estimation algorithm was generating the wrong circuit, causing the algorithm to fail for simple cases. Refer to `#9280 https://github.com/Qiskit/qiskit-terra/issues/9280` for more details.
```
Comment (Cryoris) [inline releasenotes/notes/iterative-phase-estimation-bugfix-b676ffc23cea8251.yaml:4]:
```suggestion
    Fixed a bug where :class:`.IterativePhaseEstimation` was generating the wrong circuit, causing the
    algorithm to fail for simple cases. Fixed `#9280 https://github.com/Qiskit/qiskit-terra/issues/9280`__.
```

**Comment by Cryoris** [comment]
LGTM but @jlapeyre could you also have a look? 🙂

File: releasenotes/notes/iterative-phase-estimation-bugfix-b676ffc23cea8251.yaml (hunk 1 above)
Comment (jlapeyre) [reply releasenotes/notes/iterative-phase-estimation-bugfix-b676ffc23cea8251.yaml:4]: Note that there is no newline at the end of the release note file. I'm not sure if the suggested change of @Cryoris also fixed this. But it should be fixed.

**Comment by jlapeyre** [comment]
@shifubear Thanks for sticking with this through the administrative difficulty.

It looks fine to me. You may want to consider doing this instead
```python
        qubits = [unitary.num_qubits]
        qubits.extend(range(0, unitary.num_qubits))
        qc = qc.compose(unitary_power, qubits)
```
This is a significantly faster way to construct the list. In general, I advocate using an idiomatic best way to write something. In part because the code serves as a model for further development. Which way to write it is best?
* `extend` is faster in this case
* The increase in performance won't matter in this function, because surrounding code is much slower. But it's good to habitually use the best choice. But it's not clear this is the best choice because...
* `+` is more readable and easily understandable in this case.

Unfortunately, `extend` returns `None` rather than the `list`. If it did return the `list`, then you could write
```python
qc = qc.compose(unitary_power, [unitary.num_qubits].extend(range(0, unitary.num_qubits)))
```
 and you'd have something that is both faster and no less readable. But alas this is not the case.

On balance, I favor the code as you have written in, with `+`, because the readability outweighs the performance. Do you have an opinion @Cryoris ?

**Comment by shifubear** [comment]
@jlapeyre Of course! Wanted to contribute so I'm glad there was a workaround. 

I prefer optimizing for readability in this instance, especially since the surrounding code is much slower as you mentioned. Also, my intuition is that the length of the qubits list will not be long enough for there to be a significant performance difference in the near future. 
With that said though, I was unaware of the `extend` method so I will keep note of that for future PRs!

**Comment by jlapeyre** [comment]
I don't think there is a conceivable scenario where the performance of this line would matter in this method.

**Comment by shifubear** [comment]
Hi @jlapeyre, I added the suggested changes on the release notes!

File: releasenotes/notes/iterative-phase-estimation-bugfix-b676ffc23cea8251.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+fixes:
+  - |
+    Fixed a bug where :class:`.IterativePhaseEstimation` was generating the wrong circuit, causing the
+    algorithm to fail for simple cases. Fixed `#9280 https://github.com/Qiskit/qiskit-terra/issues/9280`__.
```
Comment (jlapeyre) [inline releasenotes/notes/iterative-phase-estimation-bugfix-b676ffc23cea8251.yaml:5]: There should be a newline at the end of this file. The red circle warns about this.

Comment (shifubear) [reply releasenotes/notes/iterative-phase-estimation-bugfix-b676ffc23cea8251.yaml]: I added a newline but it looks like some tests are still failing. I can't seem to find what exactly is the cause. Do you have any pointers on how to find the error?

**Review by ElePT — COMMENTED** [review]
Hi @shifubear, you should add angle brackets to your reno hyperlink definition, sphinx expects something like: 
```
`reference_to_link_in_text <https://link.com>`__
```

**Review by Cryoris — APPROVED** [review]
LGTM thanks for the fix!

**Pull request merged**